### PR TITLE
Clean up the pixel local reconstruction code

### PR DESCRIPTION
**Pull request by fwyzard**
Address the pixel local reconstruction review comments from https://github.com/cms-sw/cmssw/pull/31721.

General clean up of the pixel local reconstruction code:
  - remove commented out and obsolete code and data members
  - use named constants more consistently
  - update variable names to follow the coding rules and for better consistency
  - use member initializer lists in the constructors
  - allow `if constexpr` in CUDA code
  - use `std::size` instead of hardcoding the array size
  - convert iterator-based loops to range-based loops
  - replace `cout` and `printf` with `LogDebug` or `LogWarning`
  - use put tokens
  - reorganise the auto-generated cfi files and use them more consistently
  - adjust code after rearranging an `#ifdef GPU_DEBUG` block
  - apply code formatting
  - other minor changes

Improve comments:
  - improve comments and remove obsolete ones
  - clarify comments and types regarding `HostProduct`
  - update comments about `GPU_SMALL_EVENTS` being kept for testing purposes
  - add notes about the original cpu code

Reuse some more common code:
  - move common pixel cluster code to `PixelClusterizerBase`
  - extend the `SiPixelCluster` constructor

Rename classes and modules for better consistency:
  - remove the `TrackingRecHit2DCUDA.h` and `gpuClusteringConstants.h` forwarding headers
  - rename `PixelRecHits` to `PixelRecHitGPUKernel`
  - rename SiPixelRecHitFromSOA to SiPixelRecHitFromCUDA
  - rename `siPixelClustersCUDAPreSplitting` to `siPixelClustersPreSplittingCUDA`
  - rename `siPixelRecHitsCUDAPreSplitting` to `siPixelRecHitsPreSplittingCUDA` 
  - rename `siPixelRecHitsLegacyPreSplitting` to `siPixelRecHitsPreSplittingLegacy`
  - rename `siPixelRecHitHostSoA` to `siPixelRecHitSoAFromLegacy`

Re-apply changes from #29805 that were lost in the Patatrack branch.

### CODE REVIEW

**Comment by fwyzard** [comment]
:construction: Validation running at fu-c2a02-37-01:/data/user/fwyzard/patatrack/validation/run_593.FAvarBWEbJ ...

**Comment by VinInn** [comment]
I do not disagree. Still some of those were just straight copies from Legacy

File: RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelDigisClustersFromSoA.cc [hunk 1]
```diff
@@ -114,8 +114,7 @@ void SiPixelDigisClustersFromSoA::produce(edm::StreamID, edm::Event& iEvent, con
         edm::LogWarning("SiPixelDigisClustersFromSoA") << "cluster below charge Threshold "
                                                        << "Layer/DetId/clusId " << layer << '/' << detId << '/' << ic
                                                        << " size/charge " << acluster.isize << '/' << acluster.charge;
-      SiPixelCluster cluster(acluster.isize, acluster.adc, acluster.x, acluster.y, acluster.xmin, acluster.ymin);
-      cluster.setOriginalId(ic);
```
Comment (VinInn) [inline RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelDigisClustersFromSoA.cc:117]: remove ```setOriginalId``` from the interface of ```SiPixelCluster```?

Comment (fwyzard) [reply RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelDigisClustersFromSoA.cc:117]: what about https://github.com/cms-patatrack/cmssw/blob/6a192beda960630de730923b10a44a5a1da5e77d/RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelClusterProducer.cc#L130 ?

File: CUDADataFormats/SiPixelCluster/interface/gpuClusteringConstants.h [hunk 2]
```diff
@@ -2,30 +2,35 @@
...
 
 namespace pixelGPUConstants {
 #ifdef GPU_SMALL_EVENTS
+  // kept for testing and debugging
   constexpr uint32_t maxNumberOfHits = 24 * 1024;
 #else
-  constexpr uint32_t maxNumberOfHits =
-      48 * 1024;  // data at pileup 50 has 18300 +/- 3500 hits; 40000 is around 6 sigma away
+  // data at pileup 50 has 18300 +/- 3500 hits; 40000 is around 6 sigma away
+  // tested on MC events with 55-75 pileup events
```
Comment (mmusich) [inline CUDADataFormats/SiPixelCluster/interface/gpuClusteringConstants.h:13]:
```suggestion
  // tested on MC events with flat 55-75 pileup events
```

File: CUDADataFormats/SiPixelCluster/interface/gpuClusteringConstants.h [hunk 3]
```diff
@@ -2,30 +2,35 @@
...
   constexpr uint32_t maxHitsInIter() { return 160; }
 #endif
   constexpr uint32_t maxHitsInModule() { return 1024; }
 
-  constexpr uint32_t MaxNumModules = 2000;
-  constexpr int32_t MaxNumClustersPerModules = maxHitsInModule();
-  constexpr uint32_t MaxHitsInModule = maxHitsInModule();  // as above
-  constexpr uint32_t MaxNumClusters = pixelGPUConstants::maxNumberOfHits;
-  constexpr uint16_t InvId = 9999;  // must be > MaxNumModules
+  constexpr uint16_t maxNumModules = 2000;
```
Comment (mmusich) [inline CUDADataFormats/SiPixelCluster/interface/gpuClusteringConstants.h:29]: why 2000? just to choose a round number?

File: CalibTracker/SiPixelESProducers/src/SiPixelGainCalibrationForHLTGPU.cc [hunk 4]
```diff
@@ -12,22 +13,21 @@ SiPixelGainCalibrationForHLTGPU::SiPixelGainCalibrationForHLTGPU(const SiPixelGa
     : gains_(&gains) {
   // bizzarre logic (looking for fist strip-det) don't ask
```
Comment (mmusich) [inline CalibTracker/SiPixelESProducers/src/SiPixelGainCalibrationForHLTGPU.cc:14]: please remove this comment

File: RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelDigisClustersFromSoA.cc (hunk 1 above)
Comment (VinInn) [reply RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelDigisClustersFromSoA.cc:117]: oops. sorry for the noise.
and cannot be done differently there...

File: RecoLocalTracker/SiPixelClusterizer/plugins/gpuCalibPixel.h [hunk 5]
```diff
@@ -4,14 +4,15 @@
 #include <cstdint>
 #include <cstdio>
 
+#include "CUDADataFormats/SiPixelCluster/interface/gpuClusteringConstants.h"
 #include "CondFormats/SiPixelObjects/interface/SiPixelGainForHLTonGPU.h"
 #include "HeterogeneousCore/CUDAUtilities/interface/cuda_assert.h"
 
 #include "gpuClusteringConstants.h"
```
Comment (mmusich) [inline RecoLocalTracker/SiPixelClusterizer/plugins/gpuCalibPixel.h:11]: twice? or are there two `gpuClusteringConstants` files?

File: RecoLocalTracker/SiPixelRecHits/plugins/SiPixelRecHitSoAFromLegacy.cc [hunk 6]
```diff
@@ -25,7 +25,7 @@
 #include "RecoLocalTracker/SiPixelRecHits/interface/PixelCPEBase.h"
 #include "RecoLocalTracker/SiPixelRecHits/interface/PixelCPEFast.h"
 
-#include "RecoLocalTracker/SiPixelRecHits/plugins/gpuPixelRecHits.h"
+#include "gpuPixelRecHits.h"
```
Comment (mmusich) [inline RecoLocalTracker/SiPixelRecHits/plugins/SiPixelRecHitSoAFromLegacy.cc:28]: for my education why is this preferred? here and and elsewhere?

File: CondFormats/SiPixelObjects/interface/SiPixelGainForHLTonGPU.h [hunk 7]
```diff
@@ -46,7 +47,7 @@ class SiPixelGainForHLTonGPU {
     assert(offset < 3088384);
```
Comment (mmusich) [inline CondFormats/SiPixelObjects/interface/SiPixelGainForHLTonGPU.h:47]: this is awful, please expand into 3088384 = 1856*416*4
Also explain what each piece stands for.

**Review by mmusich — COMMENTED** [review]
some semi-random comment / thought, while re-reading the code.

File: RecoLocalTracker/SiPixelClusterizer/plugins/gpuCalibPixel.h (hunk 5 above)
Comment (fwyzard) [reply RecoLocalTracker/SiPixelClusterizer/plugins/gpuCalibPixel.h:11]: Right...
`RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusteringConstants.h` simply includes `CUDADataFormats/SiPixelCluster/interface/gpuClusteringConstants.h`, and should be removed.

File: RecoLocalTracker/SiPixelRecHits/plugins/SiPixelRecHitSoAFromLegacy.cc (hunk 6 above)
Comment (fwyzard) [reply RecoLocalTracker/SiPixelRecHits/plugins/SiPixelRecHitSoAFromLegacy.cc:28]: The idea is that header files in a package's `interface/` directory expose that package's public interface, while header files in a package's `src/` or `plugins/` directories are private to the package.

So, whenever one sees `#include "Some/Package/plugins/header.h"` it looks like a package is including a private header from an other package.
Writing `#include "header.h"` instead makes it immediately clear that it is a header from the local package.

File: CondFormats/SiPixelObjects/interface/SiPixelGainForHLTonGPU.h (hunk 7 above)
Comment (fwyzard) [reply CondFormats/SiPixelObjects/interface/SiPixelGainForHLTonGPU.h:47]: I have no idea what each piece stands for ...

**Comment by fwyzard** [comment]
type bug-fix

File: RecoLocalTracker/SiPixelRecHits/plugins/gpuPixelRecHits.h [hunk 8]
```diff
@@ -69,8 +70,8 @@ namespace gpuPixelRecHits {
 
 #ifdef GPU_DEBUG
     if (threadIdx.x == 0) {
-      auto k = first;
-      while (digis.moduleInd(k) == InvId)
+      auto k = clusters.moduleStart(1 + blockIdx.x);
```
Comment (fwyzard) [inline RecoLocalTracker/SiPixelRecHits/plugins/gpuPixelRecHits.h:73]: @VinInn some time ago you told me how to fix this...
I didn't do it at the time, so I hope I remembered it correctly.

Comment (VinInn) [reply RecoLocalTracker/SiPixelRecHits/plugins/gpuPixelRecHits.h:73]: definitively NOT blockIdx.x: on CPU is always 1

Comment (VinInn) [reply RecoLocalTracker/SiPixelRecHits/plugins/gpuPixelRecHits.h:73]: is this code before the change in cudaCompat?

Comment (VinInn) [reply RecoLocalTracker/SiPixelRecHits/plugins/gpuPixelRecHits.h:73]: ok for auto k =

File: RecoLocalTracker/SiPixelRecHits/plugins/gpuPixelRecHits.h [hunk 9]
```diff
@@ -107,12 +105,10 @@ namespace gpuPixelRecHits {
...
       }
 
-      first += threadIdx.x;
-
       __syncthreads();
 
-      // one thead per "digi"
-
+      // one thread per "digi"
+      auto first = clusters.moduleStart(1 + blockIdx.x) + threadIdx.x;
```
Comment (VinInn) [inline RecoLocalTracker/SiPixelRecHits/plugins/gpuPixelRecHits.h:111]: and NOW it is after __syncthreads();!!!

Comment (VinInn) [reply RecoLocalTracker/SiPixelRecHits/plugins/gpuPixelRecHits.h:111]: ok, irrelevant: first is thread local

Comment (fwyzard) [reply RecoLocalTracker/SiPixelRecHits/plugins/gpuPixelRecHits.h:111]: yes - I just moved it to the first place where it was used

**Comment by fwyzard** [comment]
:construction: Validation running at fu-c2a02-37-02:/data/user/fwyzard/patatrack/validation/run_593.jFloqel2ds ...

**Comment by fwyzard** [comment]
:construction: Validation running at fu-c2a02-35-03:/data/user/fwyzard/patatrack/validation/run_593.6OCYLO1AYW ...

**Comment by fwyzard** [comment]
No impact on physics (as expected).
No regressions on throughput (as expected) when running by hand.

**Pull request merged**